File: elixir/bob/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    case {silence?(input), shouting?(input), question?(input)} do
      {true, _, _} -> "Fine. Be that way!"
      {_, true, true} -> "Calm down, I know what I'm doing!"
      {_, true, _} -> "Whoa, chill out!"
      {_, _, true} -> "Sure."
      {_, _, _} -> "Whatever."
    end
  end

  defp silence?(input) do
    String.replace(input, ~r/\s/, "") == ""
  end

  defp question?(input) do
    input
    |> String.trim_trailing()
    |> String.ends_with?("?")
  end

  defp shouting?(input) do
    cond do
      input =~ ~r/[[:lower:]]/ -> false
      input =~ ~r/[[:upper:]]/ -> true
      true -> false
    end
  end
end

File: clojure/bob/src/bob.clj
(ns bob)

(defn re-match? [re s]
  (-> (re-find re s)
      nil?
      not))

(defn yelling? [s]
  ;; contains an upper and does not contain a lower
  (and (re-match? #"\p{Lu}" s)
       (not (re-match? #"\p{Ll}" s))))

(defn question? [s]
  (re-match? #"\?\s*$" s))

(defn silent? [s]
  ;; does not contain a non-space
  (not (re-match? #"\S" s)))

(defn response-for [s]
  (cond
    (silent? s)                      "Fine. Be that way!"
    (and (yelling? s) (question? s)) "Calm down, I know what I'm doing!"
    (yelling? s)                     "Whoa, chill out!"
    (question? s)                    "Sure."
    :else                            "Whatever."))
